limiter/limiter_test.go

```diff
@@ -0,0 +1,317 @@
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 1)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 0)
+			So(l.Increment(two), ShouldBeFalse)
+			l.SetLimit("l2", 3)
+			So(l.GetLowestLimit(groups), ShouldEqual, 3)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 1)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 0)
+			So(l.Increment(two), ShouldBeFalse)
+			l.Decrement(two)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 1)
+			l.Decrement(two)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 2)
+			l.Decrement(two)
+			// at this point l2 should have been forgotten about, which means
+			// we forgot we set the limit to 3
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 2)
+			l.Decrement(two) // doesn't panic or something
+			So(l.GetLowestLimit(groups), ShouldEqual, 2)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 2)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 0)
+			So(l.Increment(two), ShouldBeFalse)
+			l.Decrement(two)
+			l.Decrement(two)
+			limits["l2"] = 3
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 3)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.GetLowestLimit(groups), ShouldEqual, 3)
+			So(l.GetRemainingCapacity(groups), ShouldEqual, 2)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.Increment(two), ShouldBeTrue)
+			So(l.Increment(two), ShouldBeFalse)
+
+			testGroup := l.vivifyGroup(groups[0])
+			So(getLowestCapacity(testGroup, -1), ShouldEqual, 3)
+		})
+
+		Convey("You can have limits of 0 and also RemoveLimit()s", func() {
+			l.SetLimit("l2", 0)
+			So(l.Increment([]string{"l2"}), ShouldBeFalse)
+
+			limits["l2"] = 0
+			l.RemoveLimit("l2")
+			So(l.Increment([]string{"l2"}), ShouldBeFalse)
+			So(l.GetLimit("l2"), ShouldEqual, 0)
+
+			limits["l2"] = -1
+			So(l.Increment([]string{"l2"}), ShouldBeFalse)
+			So(l.GetLimit("l2"), ShouldEqual, 0)
+
+			l.RemoveLimit("l2")
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.Increment([]string{"l2"}), ShouldBeTrue)
+			So(l.GetLimit("l2"), ShouldEqual, -1)
+		})
+
+		Convey("Concurrent SetLimit(), Increment() and Decrement() work", func() {
+			var incs uint64
+			var fails uint64
+			var wg sync.WaitGroup
+			for i := 0; i < 200; i++ {
+				wg.Add(1)
+				go func(i int) {
+					defer wg.Done()
+					groups := []string{"l4", "l5"}
+					if i%2 == 0 {
+						groups = []string{"l5", "l4"}
+					}
+
+					if l.Increment(groups) {
+						atomic.AddUint64(&incs, 1)
+						time.Sleep(100 * time.Millisecond)
```

Comment:
Is there a way to do this and other tests below without apparently arbitrary sleep/wait times?